### ISSUE TITLE
3.11-The complete training model

### ISSUE THREAD

**Issue by Pin-Jiun** [Pin-Jiun/Machine-Learing-HYL#26]
```py
import torch
import torchvision
from torch import nn
from torch.utils.data import DataLoader
from torch.utils.tensorboard import SummaryWriter
import time

#dataset整理data
train_set = torchvision.datasets.CIFAR10(root="./CIFAR10_dataset", train=True, transform=torchvision.transforms.ToTensor(), download=True)
test_set = torchvision.datasets.CIFAR10(root="./CIFAR10_dataset", train=False, transform=torchvision.transforms.ToTensor(), download=True)
test_size = len(test_set)

#使用dataloader跑data
train_dataloader = DataLoader(train_set, batch_size=64)
test_dataloader = DataLoader(test_set, batch_size=64)

#創建網路骨架
class PJL_NN(nn.Module):
    def __init__(self):
        super(PJL_NN,self).__init__()
        self.model = nn.Sequential(
            nn.Conv2d(3,32,5,padding=2),
            nn.MaxPool2d(2),
            nn.Conv2d(32,32,5,padding=2),
            nn.MaxPool2d(2),
            nn.Conv2d(32,64,5,padding=2),
            nn.MaxPool2d(2),
            nn.Flatten(),
            nn.Linear(64*4*4,64),
            nn.Linear(64,10)
        )

    def forward(self, x):
        x = self.model(x)
        return x





#創建網路
jiun_nn = PJL_NN()
if torch.cuda.is_available():
    print("cuda is available")
    jiun_nn = jiun_nn.cuda()

#定義loss function
loss_fn = nn.CrossEntropyLoss()
if torch.cuda.is_available():
    loss_fn = loss_fn.cuda()

#optimizer
learning_rate = 1e-2
optimizer = torch.optim.SGD(jiun_nn.parameters(), lr=learning_rate)

#開始訓練
total_train_step = 0
total_test_step = 0

#整筆資料訓練次數
epoch = 10

writer = SummaryWriter("training_log")
start_time = time.time()

for i in range(epoch):
    print("The {}-th training starts!".format(i))
    #jiun_nn.train()可寫可不寫
    for data in train_dataloader:
        imgs, targets = data
        if torch.cuda.is_available():
            imgs = imgs.cuda()
            targets = targets.cuda()
        outputs = jiun_nn(imgs)
        loss = loss_fn(outputs, targets)

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        total_train_step = total_train_step +1
        if total_train_step %100 == 0:
            end_time = time.time()
            print("Runnung time is = ", end_time-start_time)


            print("The {}-th training loss is {}".format(i,loss))
            #print("The {}-th training loss is {}".format(i,loss.item()))
            writer.add_scalar("train_loss", loss.item(), total_train_step)

    #如何知道你的測訓練集是否好? 故需要進行測試
    #使用torch.no_grad確保梯度不會進行更動到
    #jiun_nn.eval()可寫可不寫
    total_test_loss = 0
    total_accuracy = 0
    with torch.no_grad():
        for data in test_dataloader:
            imgs, targets = data
            if torch.cuda.is_available():
                imgs = imgs.cuda()
                targets = targets.cuda()
            outputs = jiun_nn(imgs)
            loss = loss_fn(outputs, targets)
            total_test_loss = total_test_loss + loss

            accuracy = (outputs.argmax(1) == targets).sum()
            total_accuracy = total_accuracy + accuracy


    print("整體測試集的loss = {}".format(total_test_loss))
    writer.add_scalar("test_loss", total_test_loss, total_test_step)
    writer.add_scalar("test_accuracy", total_accuracy/test_size, total_test_step)

    total_test_step = total_test_step + 1

    writer.close()

torch.save(jiun_nn, "jiun_nn.pth")
print("Training is done!")


```

**Comment by Pin-Jiun** [comment]
要移動到GPU的包括

1. 創建的模型
2. 數據, 包括data和label
3. loss function

可以使用to的方式移動到想要的device

```py
device = torch.device("cuda")
device = torch.device("cuda:0")
device = torch.device("cpu")

...

imgs.to(device)
```

但通常會這樣寫
```py
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
物件.to(device)
```

整體改寫

```py
import torch
import torchvision
from torch import nn
from torch.utils.data import DataLoader
from torch.utils.tensorboard import SummaryWriter
import time

if torch.cuda.is_available():
    print("cuda is available")


#dataset整理data
train_set = torchvision.datasets.CIFAR10(root="./CIFAR10_dataset", train=True, transform=torchvision.transforms.ToTensor(), download=True)
test_set = torchvision.datasets.CIFAR10(root="./CIFAR10_dataset", train=False, transform=torchvision.transforms.ToTensor(), download=True)
test_size = len(test_set)

#使用dataloader跑data
train_dataloader = DataLoader(train_set, batch_size=64)
test_dataloader = DataLoader(test_set, batch_size=64)

#創建網路骨架
class PJL_NN(nn.Module):
    def __init__(self):
        super(PJL_NN,self).__init__()
        self.model = nn.Sequential(
            nn.Conv2d(3,32,5,padding=2),
            nn.MaxPool2d(2),
            nn.Conv2d(32,32,5,padding=2),
            nn.MaxPool2d(2),
            nn.Conv2d(32,64,5,padding=2),
            nn.MaxPool2d(2),
            nn.Flatten(),
            nn.Linear(64*4*4,64),
            nn.Linear(64,10)
        )

    def forward(self, x):
        x = self.model(x)
        return x

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")


#創建網路
jiun_nn = PJL_NN()
jiun_nn = jiun_nn.to(device)

#定義loss function
loss_fn = nn.CrossEntropyLoss()
loss_fn = loss_fn.to(device)

#optimizer
learning_rate = 1e-2
optimizer = torch.optim.SGD(jiun_nn.parameters(), lr=learning_rate)

#開始訓練
total_train_step = 0
total_test_step = 0

#整筆資料訓練次數
epoch = 10

writer = SummaryWriter("training_log")
start_time = time.time()

for i in range(epoch):
    print("The {}-th training starts!".format(i))
    #jiun_nn.train()可寫可不寫
    for data in train_dataloader:
        imgs, targets = data
        imgs = imgs.to(device)
        targets = targets.to(device)
        outputs = jiun_nn(imgs)
        loss = loss_fn(outputs, targets)

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        total_train_step = total_train_step +1
        if total_train_step %100 == 0:
            end_time = time.time()
            print("Runnung time is = ", end_time-start_time)


            print("The {}-th training loss is {}".format(i,loss))
            #print("The {}-th training loss is {}".format(i,loss.item()))
            writer.add_scalar("train_loss", loss.item(), total_train_step)

    #如何知道你的測訓練集是否好? 故需要進行測試
    #使用torch.no_grad確保梯度不會進行更動到
    #jiun_nn.eval()可寫可不寫
    total_test_loss = 0
    total_accuracy = 0
    with torch.no_grad():
        for data in test_dataloader:
            imgs, targets = data
            imgs = imgs.to(device)
            targets = targets.to(device)
            outputs = jiun_nn(imgs)
            loss = loss_fn(outputs, targets)
            total_test_loss = total_test_loss + loss

            accuracy = (outputs.argmax(1) == targets).sum()
            total_accuracy = total_accuracy + accuracy


    print("整體測試集的loss = {}".format(total_test_loss))
    writer.add_scalar("test_loss", total_test_loss, total_test_step)
    writer.add_scalar("test_accuracy", total_accuracy/test_size, total_test_step)

    total_test_step = total_test_step + 1

    writer.close()

torch.save(jiun_nn, "jiun_nn.pth")
print("Training is done!")


```

**Comment by Pin-Jiun** [comment]
### 實際預測

將訓練好的model下載下來進行驗證

```py
import torch
import torchvision
from torch import nn
from PIL import Image

image_path = "dog.jpg"
image = Image.open(image_path)

transform = torchvision.transforms.Compose([
    torchvision.transforms.Resize((32,32)),
    torchvision.transforms.ToTensor(),
])

image = transform(image)
image = torch.reshape(image, (1,3,32,32))

class PJL_NN(nn.Module):
    def __init__(self):
        super(PJL_NN,self).__init__()
        self.model = nn.Sequential(
            nn.Conv2d(3,32,5,padding=2),
            nn.MaxPool2d(2),
            nn.Conv2d(32,32,5,padding=2),
            nn.MaxPool2d(2),
            nn.Conv2d(32,64,5,padding=2),
            nn.MaxPool2d(2),
            nn.Flatten(),
            nn.Linear(64*4*4,64),
            nn.Linear(64,10)
        )

    def forward(self, x):
        x = self.model(x)
        return x


model = torch.load("jiun_nn.pth", map_location=torch.device('cpu'))
# print(model)

model.eval()
with torch.no_grad():
    output = model(image)

print(output.argmax(1))

```

https://www.cs.toronto.edu/~kriz/cifar.html

**Comment by Pin-Jiun** [comment]
在 PyTorch 中，model.train() 是一個用於設置模型處於訓練模式的方法。這個方法主要用於啟用具有訓練行為的模型組件，例如 dropout 或 batch normalization。當你訓練模型時，你應該在每個 epoch 或每個 mini-batch 開始時調用 model.train()，以確保模型處於訓練模式。

舉例來說，當模型中包含 dropout 層時，在訓練模式下，dropout 層會在前向傳播中保留隨機丟棄的神經元。而在測試/評估模式下，dropout 層通常應該被關閉，以保證一致的預測結果。

```PY

model = YourModel()
optimizer = torch.optim.SGD(model.parameters(), lr=0.01)
criterion = torch.nn.CrossEntropyLoss()

# 訓練迴圈
for epoch in range(num_epochs):
    model.train()  # 啟用訓練模式
    for inputs, labels in train_dataloader:
        optimizer.zero_grad()
        outputs = model(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()

    # 在每個 epoch 結束後，你可能會進行模型的評估，這時你可以使用 model.eval()
    model.eval()  # 啟用評估模式
    with torch.no_grad():
        # 驗證或測試的程式碼


```
